Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt

```diff
@@ -0,0 +1,192 @@
+        val json = JSONObject()
+        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault())
+
+        for (param in urlQueryParameters.values) {
+            val paramDict = JSONObject()
+            paramDict.put("name", param.name)
+            paramDict.put("value", param.value ?: JSONObject.NULL)
+            paramDict.put("timestamp", param.timestamp?.let { dateFormat.format(it) })
+            paramDict.put("isDeeplink", param.isDeepLink)
+            paramDict.put("validityWindow", param.validityWindow)
+
+            json.put(param.name.toString(), paramDict)
+        }
+
+        return json
+    }
+
+    @VisibleForTesting
+    internal fun deserializeFromJson(json: JSONObject): MutableMap<String, BranchUrlQueryParameter> {
+        val result = mutableMapOf<String, BranchUrlQueryParameter>()
+        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault())
+
+        val keys = json.keys()
+        while (keys.hasNext()) {
+            val key = keys.next()
+            val temp = json.getJSONObject(key)
+
+            val param = BranchUrlQueryParameter()
+            param.name = temp.getString("name")
+
+            if (!temp.isNull("value")) {
+                param.value = temp.getString("value")
+            }
+
+            param.timestamp = temp.getString("timestamp")?.let { dateFormat.parse(it) }
+            param.validityWindow = temp.getLong("validityWindow")
+
+            if (!temp.isNull("isDeeplink")) {
+                param.isDeepLink = temp.getBoolean("isDeeplink")
+            } else {
+                param.isDeepLink = false
+            }
+
+            param.name?.let { paramName ->
+                result[paramName] = param
+            }
+        }
+
+        return result
+    }
+
+    @VisibleForTesting
+    internal fun checkForAndMigrateOldGclid() {
+        //Check if there is an existing Gclid, validityWindow, etc. If there is, create a new BranchUrlQueryParameter for it.
+        val newGclid = urlQueryParameters[Defines.Jsonkey.Gclid.key]
+        if (newGclid?.value == null) {
+            val existingGclidValue = prefHelper.referrerGclid
+            if (existingGclidValue != null && existingGclidValue != PrefHelper.NO_STRING_VALUE) {
+                val existingGclidValidityWindow = prefHelper.referrerGclidValidForWindow
+
+                val gclid = BranchUrlQueryParameter(
+                    name = Defines.Jsonkey.Gclid.key,
+                    value = existingGclidValue,
+                    timestamp = Date(),
+                    validityWindow = existingGclidValidityWindow,
+                    isDeepLink = false
+                )
+
+                urlQueryParameters[Defines.Jsonkey.Gclid.key] = gclid
+
+                prefHelper.setReferringUrlQueryParameters(serializeToJson(urlQueryParameters))
+                prefHelper.referrerGclid = null
```

Comment:
I think there was a whole `JSONObject` that encompassed both sub JSONObjects, would it make sense to null that out instead?